File: blind-75/stacks/valid-parentheses/typescript/valid-parentheses.test.ts
import { expect, test } from "bun:test";
import { validParentheses } from "./valid-parentheses";

test("Example 1", () => {
  expect(validParentheses("[]")).toBe(true);
});

test("Example 2", () => {
  expect(validParentheses("([{}])")).toBe(true);
});

test("Example 3", () => {
  expect(validParentheses("[(])")).toBe(false);
});

test("empty string", () => {
  expect(validParentheses("")).toBe(true);
});

test("single opener", () => {
  expect(validParentheses("(")).toBe(false);
});

test("single closer", () => {
  expect(validParentheses(")")).toBe(false);
  expect(validParentheses("]")).toBe(false);
  expect(validParentheses("}")).toBe(false);
});

test("closer first", () => {
  expect(validParentheses(")(")).toBe(false);
});

test("all closers", () => {
  expect(validParentheses("))]]")).toBe(false);
});

test("mismatched pair", () => {
  expect(validParentheses("(]")).toBe(false);
});

test("interleaved", () => {
  expect(validParentheses("([)]")).toBe(false);
});

test("closer with empty stack", () => {
  expect(validParentheses("())(()")).toBe(false);
});

test("unclosed openers", () => {
  expect(validParentheses("((")).toBe(false);
  expect(validParentheses("([](")).toBe(false);
});

test("sequential valid", () => {
  expect(validParentheses("()[]{}")).toBe(true);
});

test("deeply nested valid", () => {
  expect(validParentheses("((([[[{{{}}}]]])))")).toBe(true);
});

test("long repeated valid", () => {
  expect(validParentheses("()".repeat(5000))).toBe(true);
});

File: blind-75/stacks/valid-parentheses/go/valid_parentheses_test.go
package validparentheses

import (
	"strings"
	"testing"
)

func TestValidParentheses(t *testing.T) {
	tests := []struct {
		name	string
		s			string
		want	bool
	} {
		{"Example 1", "[]", true},
		{"Example 2", "([{}])", true},
		{"Example 3", "[(])", false},
		{"Empty string", "", true},
		{"Single opener", "(", false},
		{"Single closer )", ")", false},
		{"Single closer ]", "]", false},
		{"Single closer }", "}", false},
		{"Closer first", ")(", false},
		{"All closers", "))]]", false},
		{"Mismatched pair", "(]", false},
		{"Interleaved", "([)]", false},
		{"Closer with empty stack", "())(()", false},
		{"Unclosed openers", "((", false},
		{"Unclosed openers with prefix", "([](", false},
		{"Sequential valid", "()[]{}", true},
		{"Deeply nested valid", "((([[[{{{}}}]]])))", true},
		{"Long repeated valid", strings.Repeat("()", 5000), true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if got := ValidParentheses(tt.s); got != tt.want {
				t.Errorf("ValidParentheses(%v) = %v want %v", tt.s, got, tt.want)
			}
		})
	}
}